src/render.rs

```diff
@@ -0,0 +1,301 @@
+use crate::sprites::get_player_sprite_4;
+use crate::types::{
+    Coordinate, Entity, GameWindow, GlyphCell, GlyphCells, RenderGraph, RenderItem,
+    RenderNode, World, Zoom,
+};
+
+use ndarray::Array2;
+use ratatui::style::Color;
+use ratatui::{buffer::Buffer, layout::Rect, widgets::Widget};
+
+pub fn is_in_bounds<T>(position: &Coordinate, buffer: &Array2<T>) -> bool {
+    let (height, width) = buffer.dim(); // Get the dimensions of the buffer
+
+    position.x < width || position.y < height
+}
+
+fn render_pixels(
+    pixel_size: usize,
+    item: RenderItem,
+    glyph_buffer: GlyphCells,
+) -> GlyphCells {
+    match item {
+        RenderItem::Board(board) => {
+            let mut glyph_buffer = glyph_buffer.clone();
+
+            for (yi, row) in board.rows().into_iter().enumerate() {
+                for (xi, tile) in row.iter().enumerate() {
+                    if let Some(color) = tile.color() {
+                        let pos = Coordinate { x: xi, y: yi };
+                        glyph_buffer =
+                            draw_square(color, pos, pixel_size, glyph_buffer.clone());
+                    }
+                }
+            }
+            glyph_buffer
+        }
+        RenderItem::Entity(entity) => {
+            let pos = entity.get_position();
+            draw_square(entity.color(), pos, pixel_size, glyph_buffer)
+        }
+    }
+}
+
```

Comment:
I tried something like:

```
fn draw_square(
    color: Color,
    position: Coordinate,
    size: usize,
    glyph_buffer: GlyphCells,
) -> GlyphCells {
    let pos = Coordinate {
        x: position.x * size,
        y: position.y / 2 * size,
    };
    let mut glyph_buffer = glyph_buffer.clone();
    let mut slice =
        glyph_buffer.slice_mut(s![pos.y..pos.y + size, pos.x..pos.x + size]);
    slice.map_mut(|cell| {
        cell.glyph = '▀';
        cell.fg = Some(color);
        cell.bg = Some(color);
    });
    glyph_buffer
}
```

Which I like the use of a view and simpler iteration, the issue is that because each real pixel is half an index drawing a single row of pixels doesn't work easily. I'm trying to find a clean approach that works like this where if the size of the box to draw is an odd number the last row of the slice only fills in `row.fg`.

Reply:
@sabjorn Any input on this conundrum? I really like the idea of just operating only on the slice that needs to update, but indexing might just be fundamentally easier to read and write for this transform.

Second best idea I have is zipping Option<Colors>, and then zipping that together with mutable cells. I think it's guaranteed to be less readable than the for loops tho.

Reply:
A structure that gets zipped to look like:

```
[((Some<Color>, Some<Color>), &mut GlyphCell),...].for_each(|cell| ... )
```